baseclasses/SolverHist.py

```diff
@@ -0,0 +1,210 @@
+"""
+==============================================================================
+BaseClasses: Solver History Class
+==============================================================================
+@File    :   SolverHist.py
+@Date    :   2020/11/30
+@Author  :   Alasdair Christison Gray
+@Description :
+"""
+
+# ==============================================================================
+# Standard Python modules
+# ==============================================================================
+from warnings import warn
+import copy
+import os
+
+# ==============================================================================
+# External Python modules
+# ==============================================================================
+import numpy as np
+
+# ==============================================================================
+# Extension modules
+# ==============================================================================
+
+
+class SolverHist(object):
+    """
+    The SolverHist class stores various useful values during a nonlinear solution in a nested dictionary structure, by
+    default, the history only stores the values from the latest iteration which are used for printing to the terminal.
+    Storing of the full solution history is enabled using the 'writeConvergenceHist' option
+    """
+
+    def __init__(self, options):
+        """
+        Create an empty history dictionary based on the current solver options
+        """
+
+        if options["writeconvergencehist"]:
+            length = int(options["maxiter"] + 1)
+            self.storeValues = True
+        else:
+            length = 1
+            self.storeValues = False
```

Comment:
Is this meant to replicate the print behavior of e.g. ADflow in non-linear TACS? In the long run we might want to have a consistent approach to all solvers but that would imply backwards incompatibilities and more general considerations on what we want to be printed (ADflow is pretty mature there, we probably want to maintain the same capabilities?)

Reply:
Exactly, my intention with the printing part of this is to more or less recreate ADflow's capabilities in a way that we can use in all our solvers.

Reply:
I don't think we should expect to make ADflow use this, since all the data is in Fortran and it doesn't make sense to pass them to Python just to print stuff.

Reply:
Right, for ADflow we can probably continue using the Fortran printing stuff and just use this class for history storage. I was meaning that we can use this class to replicate ADflow's printing capabilities in other solvers